Cargo.toml

```
@@ -87,7 +87,7 @@ ctrlc = { version = "3.4.4", optional = true }
 hostname = { version = "0.4.0", optional = true }
 libffi = { version = "3.2.0", optional = true }
 native-tls = { version = "0.2.12", optional = true }
-reqwest = { version = "0.11.27", optional = true }
+reqwest = { version = "0.11.*", optional = true }
```

I think its generally more common to specify `0` rather than `*` 
```suggestion
reqwest = { version = "0.11.0", optional = true }
```

---

Another option would be to just leave of the patch segment i.e. to specify `0.11`

---

> I think its generally more common to specify 0 rather than *

But that would fix it to `0.11.0` wouldn't it? The intent here is to make it flexible so that it can match whatever version Deno uses. `0.11` is better in this case I think.

---

[Cargo treats `X.Y.Z` as `^X.Y.Z`](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#caret-requirements) which is equivalent to `>= X.Y.Z, < X+1.0.0`

---

So `0.11.0` , `0.11.*` and `0.11` are all equivalent

---

Of these variants, `0.11.*` seems to best indicate that we did it this way for this reason.

---

I think I would prefer a comment with an explanation over using `*`, but it doesn't really matter in the end.

---

Added a comment. Wondering if `*` would be a good default for patch versions in general, unless there is really a reason to require a specific patch level.

---

Cargo advises is to 

> Avoid constraining the upper bound of a version to be anything less than the next semver incompatible version (e.g. avoid ">=2.0, <2.4") as other packages in the dependency tree may require a newer version, leading to an unresolvable error (see [#9029](https://github.com/rust-lang/cargo/issues/9029)). Consider whether controlling the version in your [Cargo.lock](https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html) would be more appropriate.

Which the old deno version does not follow causing your issue.

Not that this changes anything, but I think deno is in the wrong here. 
- The problem is an increased startup time which is no a critical bug. 
- deno is a libray crate and as such should have left this up to downstream binary creates whether this warrants pinning the dependency, maybe they would even like to use the feature, so they would have to pay the price anyways, but are now locked by deno from upgrading to a version that has it

Also this doesn't work for all dependencies as below `< 1.0.0` the path version acts as the minor version (i.e. new features require just a patch increase not a minor version increase) which means downgrading might remove features and as such break code. [Also some authors don't believe in the usefulness of distinguishing minor and patch version](https://github.com/rust-lang/cc-rs/issues/601) and always keep the minor version at 0.

Also I think the added comment is too vague. Given the cargo advise listed above there shouldn't be a too high patch version as there shouldn't be an upper bound limiting us to just use the highest version. So I think the explicit instance (mentioning deno, this Issue/PR, https://github.com/denoland/deno_core/issues/403, possibly https://github.com/seanmonstar/reqwest/pull/1955#issuecomment-1847022773) should be named, as it is an unexpected problem to have in the first place.

Also just downgrading without checking is bad as you might be missing critical bug fixes.
(there exists the unstable [--Zminimal-version](https://github.com/rust-lang/cargo/issues/5657) wich one could use in CI to check this)
Generally I try to keep my Cargo.toml and Cargo.lock in sync as the version in Cargo.lock is the version that was tested (using `cargo-edit`s cargo upgrade command, which is in the process of being incorporated into cargo itself, https://github.com/rust-lang/cargo/issues/12425, https://github.com/rust-lang/cargo/issues/10498). A lower version might work or it might not.

---

Newer versions of Deno have upgraded to reqwest 0.12.x. But Deno and Holochain have a couple of overlapping dependencies, which forces us to either pick matching versions, or create forks. Upgrading either of the two is a bit of an ordeal, which we do periodically, mostly driven by Holochain updates (because those tend to make big differences to the performance of our build).

So, that old version of Deno (that we're stuck with until we do the next batch of updates) has a reason to pin reqwest.

If there is no strong reason for Scryer to constrain the version, it might make sense not to do so. It would make it much easier for us to keep track of the other PRs going on in Scryer atm, esp. around lib_machine which would be great for us to test against our full integration.

Please let me know if there is something that I can do to get this PR in - or: if you think this shouldn't get merged.

---

I know that Scryer is mainly build to an executable and in that case there is no reason to not be specific about versions down to the patch level.

But when used as a library (which is what we introduced with lib_machine and our use-case) I think it's best to be flexible when possible. We don't have that problem now with other deps, but it might arise. Leading me to think that for any crate that's used as a lib, to not specify patch level version numbers if not absolutely required.

The Cargo.lock file would actually be a way to have builds of the Scryer executable use fixated and controlled versions of all deps, which would not affect the library use-case, since lock files are only build and used in the context of the consuming root crate within a build.

---

My comment wasn't a rejection of this PR in and of itself. 
I think your problem/use-case provides a fair reason for lowering the requirement in this specific case.
It was just a disagreement with the suggestion to stop constraining patch version in general.
I think the comment should be more specific for why we reduced the version requirement e.g.
```suggestion
# the version requirement of reqwest is kept low for compatibility with old deno versions 
# that pin request to 0.11.20
reqwest = { version = "0.11.0", optional = true }
```